context/src/main/java/io/opentelemetry/context/ThreadLocalContextStorage.java

```diff
@@ -30,14 +31,30 @@ public Scope attach(Context toAttach) {
 
     THREAD_LOCAL_STORAGE.set(toAttach);
 
-    return () -> {
-      if (current() != toAttach) {
+    return new ScopeImpl(beforeAttach, toAttach);
+  }
+
+  private class ScopeImpl implements Scope {
+    @Nullable private final Context beforeAttach;
+    private final Context toAttach;
+    private boolean closed;
+
+    private ScopeImpl(@Nullable Context beforeAttach, @Nonnull Context toAttach) {
+      this.beforeAttach = beforeAttach;
+      this.toAttach = toAttach;
+    }
+
+    @Override
+    public void close() {
+      if (!closed && current() == toAttach) {
```

Comment:
Does `closed` need to be `AtomicBoolean` with `compareAndSet`?

Reply:
My way of thinking is the following:
1. In general, scopes should only be closed on the same thread they were started on, so the whole scenario is single-threaded. And if not, then there are two other things to consider.
2. if the scope is closed, the second condition `current() == toAttach` should be false
3. with high concurrency, we can be in the  `!closed && current() == toAttach` branch more than once, but then we'd set the same result (`beforeAttach`), so it should be ok

so unless I'm missing something, we don't need atomic operations here and it's nice not to add any new overhead.

Reply:
That reasoning makes sense to me 👍